src/ray/gcs/gcs_server/gcs_resource_manager.h

```diff
@@ -152,6 +166,77 @@ class GcsResourceManager : public rpc::NodeResourceInfoHandler,
                          std::unordered_map<google::protobuf::Map<std::string, double>,
                                             rpc::ResourceDemand> *aggregate_load);
 
+  /// @brief  Add a new node state.
+  /// @param instance_id Instance id that's associated with the cloud provider if
+  /// relevant.
+  /// @param node_id Node id.
+  /// @param version Current cluster state version.
+  void AddNewNodeState(const std::string &instance_id,
+                       const std::string &node_id,
+                       int64_t version);
+
+  /// @brief  Remove a node state.
+  /// @param node_id Node id.
+  void RemoveNodeState(const NodeID &node_id);
+
+  /// @brief Get the current cluster resource state version.
+  ///
+  /// When the cluster state (that's currently reported to autoscaler) that represents the
+  /// resource state of the cluster is updated, a monotonically increasing version number
+  /// will be incremented. Current cluster state consists of :
+  ///   - node membership: when node is added/removed/drained.
+  ///   - cluster resources: available/total resources of the cluster.
+  ///   - resource requests: e.g. placement group requests, tasks/actors resource requests
+  ///     or cluster resources requested explicitly.
+  ///
+  /// Right now, consecutive versions are NOT guaranteed to be different. The version is
```

Comment:
> consecutive versions are NOT guaranteed to be different.

is it something everyone agrees on? Feel like this defeats the purpose of "version"?

Reply:
> is it something everyone agrees on?

Maybe not

> Feel like this defeats the purpose of "version"?

I think it doesn't really make a big difference? I think the version has 2 main purposes: 
1. Detect stale info -> so as long as it increments monotonically with every change. 
2. Help debug  -> given a version, one could trace the state. 

And not sure the requirement of having consecutive versions have to be different yields any benefits.

Reply:
Yeah I guess without the guarantee 2 is possible, and then I am okay with that. but let's wait for @scv119's feedback.

Reply:
i think ideally we want to use version to detect change too. is it very hard to achieve?

Reply:
> Is it very hard to achieve? 

I think with the current code - where we just do overwrite w/o knowing if it actually changes (e.g. placement group load), we will prob have to assume there is a delta.  But at other places, we just need to be careful when we bump the version (only if there's data changed).